plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,833 @@
+            _Ex = np.nan_to_num(_Ex, nan=0.0 * u.V / u.m)
+            _Ey = np.nan_to_num(_Ey, nan=0.0 * u.V / u.m)
+            _Ez = np.nan_to_num(_Ez, nan=0.0 * u.V / u.m)
+            _Bx = np.nan_to_num(_Bx, nan=0.0 * u.T)
+            _By = np.nan_to_num(_By, nan=0.0 * u.T)
+            _Bz = np.nan_to_num(_Bz, nan=0.0 * u.T)
+
+            # Add the values interpolated for this grid to the totals
+            Ex += _Ex
+            Ey += _Ey
+            Ez += _Ez
+            Bx += _Bx
+            By += _By
+            Bz += _Bz
+
+        # Create arrays of E and B as required by push algorithm
+        E = np.array(
+            [Ex.to(u.V / u.m).value, Ey.to(u.V / u.m).value, Ez.to(u.V / u.m).value]
+        )
+        E = np.moveaxis(E, 0, -1)
+        B = np.array([Bx.to(u.T).value, By.to(u.T).value, Bz.to(u.T).value])
+        B = np.moveaxis(B, 0, -1)
+
+        # Calculate the adaptive timestep from the fields currently experienced
+        # by the particles
+        # If user sets dt explicitly, that's handled in _adaptive_dt
+        dt = self._adaptive_dt(Ex, Ey, Ez, Bx, By, Bz)
+
+        # TODO: Test v/c and implement relativistic Boris push when required
+        # vc = np.max(v)/_c
+
+        if self.is_uniform_time:
+            # If a uniform timestep is specified, use that
+            self.time += dt
+        else:
+            # If dt is not a scalar (i.e. uniform time), make sure it can be multiplied by a
+            # [nparticles, 3] shape field array
+            dt = dt[tracked_mask, np.newaxis]
+
+            # Increment the tracked particles' time by dt
+            self.time[tracked_mask] += dt
+
+        self.x[tracked_mask], self.v[tracked_mask] = boris_push(
+            pos_tracked, vel_tracked, B, E, self.q, self.m, dt, inplace=False
+        )
+
+    @property
+    def on_any_grid(self):
+        """
+        Binary array for each particle indicating whether it is currently
+        on ANY grid.
+        """
+        return np.sum(self.on_grid, axis=-1) > 0
+
+    @property
+    def vmax(self):
+        """
+        Calculate the maximum velocity.
+        Used for determining the grid crossing maximum timestep.
+        """
+        tracked_mask = self._tracked_particle_mask()
+
+        return np.max(np.linalg.norm(self.v[tracked_mask], axis=-1))
+
+    def _validate_run_inputs(
+        self, stop_condition, save_routine, dt, field_weighting: str
+    ):
+        if not isinstance(stop_condition, AbstractStopCondition):
+            raise TypeError("Please specify a valid stop condition.")
+
+        if not isinstance(save_routine, AbstractSaveRoutine):
+            raise TypeError("Please specify a valid save routine")
+
+        # Will the simulation follow a uniform time step?
+        # This must be the case for certain stopping conditions and saving routines
+        self.is_uniform_time = isinstance(dt, u.Quantity) and isinstance(
```

Comment:
There are three cases to consider here: 

1) Adaptive timestep (nonuniform) computed separately for every particle , dt=[nparticles,]
2) Adaptive timestep (nonuniform)  but all particles use the same one (the min. of those computed for all particles). This needs to be added as a keyword option somehow, and you should be able to use save functions on this? 
3) Uniformly spaced timesteps (behavior you get if you just set dt to a fixed value).

Reply:
I have been using "uniform" to mean that the simulation uses the same time step for all of the particles. Is this term accurate? Or should I switch to something a little less ambiguous.. 

Either way, I fail to account for the case of a uniform adaptive time step, something that should definitely be implemented.

Reply:
I think it's accurate but potential ambiguous, since it could be understood in either way. Maybe the best way to handle this is just by raising appropriate errors if the timestep (stored as dt in the Tracker object) doesn't match the requirements of the save condition?

Reply:
Is this what you were looking for?

https://github.com/PlasmaPy/PlasmaPy/blob/61630328929ae6e8ac09bec9c0517e9f4ed07033/plasmapy/simulation/particle_tracker.py#L681

An error is raised if we require a uniform (over all particles) time step and one is not explicitly set. The adaptive time step is used otherwise

Reply:
Yes, I think this is good. Maybe to solve the ambiguity we can introduce two class properties 

`self.is_adaptive_timestep` -> This property, says whether or not an adaptive timestep is being used. I think this is a little more clear? 

`self.is_syncronized_timesteps` -> Timesteps are the same for each particle, or different for each particle. 

At least right now, if `self.is_adaptive_timestep=False`, then `self.is_syncronized_timesteps=False`. But if you are using the adaptive timestep, you could either be applying one timestep to all particles (`self.is_syncronized_timesteps=True`) or you could be calculating a different one for each particle (`self.is_syncronized_timesteps=False`) which is what the radiography tracer does right now by default (since the particles are not interacting). 

Making these class properties will make it easier to document clearly what each means in the docstring.